doc/serdi.1
.TH SERDI 1 "06 Jan 2019"

.SH NAME
.B serdi \- Read and write RDF syntax

.SH SYNOPSIS
serdi [OPTION]... INPUT BASE_URI

.SH OPTIONS

.TP
\fB\-a\fR
Write ASCII output if possible.

.TP
\fB\-b\fR
Fast bulk output for large serialisations.

.TP
\fB\-c PREFIX\fR
Chop PREFIX from matching blank node IDs.

.TP
\fB\-e\fR
Eat input one character at a time, rather than a page at a time which is the
default.  This is useful when reading from a pipe since output will be
generated immediately as input arrives, rather than waiting until an entire
page of input has arrived.  With this option serdi uses one page less memory,
but will likely be significantly slower.

.TP
\fB\-f\fR
Fast serialisation without object inlining.  This affects only model
(non-streaming) serialisation, and disables searching and statement reordering
for pretty printing.  Statements will be written in simple sorted order, which
is faster, but may result in less readable output in Turtle or TriG.

.TP
\fB\-h\fR
Print the command line options.

.TP
\fB\-i SYNTAX\fR
Read input as SYNTAX.
Valid values (case-insensitive): turtle, ntriples, trig, nquads.

.TP
\fB\-k BYTES\fR
Parser stack size.

.TP
\fB\-l\fR
Lax (non-strict) parsing.

.TP
\fB\-m\fR
Build and serialise a model.  This loads the complete input data into memory
before writing the output.  This will normalize statement order, at the cost of
performance and memory consumption.

.TP
\fB\-o SYNTAX\fR
Write output as SYNTAX.
Valid values (case-insensitive): turtle, ntriples, trig, nquads.

.TP
\fB\-p PREFIX\fR
Add PREFIX to blank node IDs.

.TP
\fB\-q\fR
Suppress all output except data.

.TP
\fB\-r ROOT_URI\fR
Keep relative URIs within ROOT_URI.

.TP
\fB\-s INPUT\fR
Parse INPUT as a string (terminates options).

.TP
\fB\-t\fR
Write terser output without newlines.

.TP
\fB\-v\fR
Display version information and exit.

.SH AUTHOR
Serdi was written by David Robillard <d@drobilla.net>

.SH COPYRIGHT
Copyright \(co 2011-2019 David Robillard.
.br
License: <http://www.opensource.org/licenses/isc>
.br
This is free software; you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.

.SH "SEE ALSO"
<http://drobilla.net/software/serd>


NEWS
serd (1.0.0) unstable;

  * Add SerdBuffer for mutable buffers to keep SerdChunk const-correct
  * Remove useless character counting from API
  * Use char* for strings in public API
  * Rename SerdChunk to SerdStringView
  * Make serd_strtod API const-correct
  * Make nodes opaque
  * Remove half-baked serd_uri_to_path()
  * Bring read/write interface closer to C standard
  * Add SerdWorld for shared library state
  * Use a fixed-size reader stack
  * Simplify writer style options
  * Simplify streaming API and improve pretty printing
  * Add logging functions to public API
  * Add model for storing statements in memory
  * Add support for validation
  * Add option for writing terse output without newlines
  * Add support for writing terse collections

 -- David Robillard <d@drobilla.net>  Sun, 10 Mar 2019 18:09:18 +0100

serd (0.30.1) unstable;

  * Fix GCC 4 build
  * Fix resolving some URIs against base URIs with no trailing slash
  * Fix colliding blank nodes when parsing TriG
  * Fix missing parse error messages
  * Fix parsing TriG graphs with several squashed trailing dots
  * Fix parsing "a" abbreviation without padding whitespace
  * Improve documentation

 -- David Robillard <d@drobilla.net>  Sat, 30 Mar 2019 12:57:54 +0100

serd (0.30.0) stable;

  * Upgrade to waf 2.0.9 and autowaf a328983
  * Fix length of base64 encoded nodes
  * Clarify errors returned by serd_env_expand()
  * Fix reported error when reading statements with only a blank node
  * Fix parsing local names that end with escaped dots
  * Improve relative URI serialisation
  * Add serdi option to write ASCII output
  * Make serdi guess input syntax from extension if unspecified
  * Make serdi syntax options case-insensitive
  * Add basic error tolerant parsing for line-based formats
  * Support strict parsing of prefixed names
  * Report errors for all failing test cases
  * Use more efficient stack allocation and growth policies
  * Add NTriples test suite
  * Fix building with MSVC
  * Clean up testing code
  * Add serd_free for freeing memory allocated by serd

 -- David Robillard <d@drobilla.net>  Sun, 22 Jul 2018 20:34:09 +0200

serd (0.28.0) stable;

  * Add support for reading from a user provided callback
  * Add serd_node_from_substring()
  * Fix strict parsing of abolute URI schemes
  * Fix parsing of hex escapes in file URIs (thanks Johannes Mueller)
  * Gracefully handle applications that write corrupt UTF-8
  * Fix hangs when reading corrupt UTF-8

 -- David Robillard <d@drobilla.net>  Fri, 21 Jul 2017 08:28:47 +0200

serd (0.26.0) stable;

  * Add support for TriG and NQuads
  * Add support for braindead SPARQL PREFIX and BASE directives

 -- David Robillard <d@drobilla.net>  Mon, 09 Jan 2017 20:10:35 -0500

serd (0.24.0) stable;

  * Add serd_node_new_relative_uri()
  * Fix construction and comparison of URIs with UTF-8 characters
  * Report I/O errors with message and return appropriate status code
  * Report missing statement separator errors
  * Fix potential out of bounds read
  * Fix unaligned memory access, undefined behaviour which breaks on ARM
  * Fix documentation generation
  * Update serdi man page

 -- David Robillard <d@drobilla.net>  Mon, 19 Sep 2016 22:21:47 -0400

serd (0.22.0) stable;

  * Remove dependence on fmax() to avoid portability issues
  * Fix serd_reader_read_file() for URIs with escaped characters (spaces)
  * Add serd_reader_set_strict() and -l (lax) option to serdi to tolerate
    parsing URIs with escaped characters
  * Fix reading statements ending with a blank then dot with no space
  * Fix clash resolution when a blank node ID prefix is set
  * Fix serializing fractional decimals that would round up
  * Add support for Turtle named inline nodes extension
  * Report errors for invalid IRI characters and missing terminators
  * Show serdi errors in standard format
  * Fix warnings when building with ISO C++ compilers
  * Upgrade to waf 1.8.14

 -- David Robillard <d@drobilla.net>  Thu, 08 Oct 2015 15:34:18 -0400

serd (0.20.0) stable;

  * Support new RDF 1.1 Turtle
  * Don't write xsd:decimal literals to Turtle bare if they would not be read
    back with the same type
  * Fix possible crash in serd_writer_end_anon() when writing invalid lists
  * Generate blank names like _:b1 and _:B2 not _:genid1 _:docid2
  * Correctly handle posix_memalign failure
  * Fix const-correctness violation for reader input string
  * Add -lm to pkg-config libs
  * Update to waf 1.7.14

 -- David Robillard <d@drobilla.net>  Fri, 08 Aug 2014 18:02:54 -0400

serd (0.18.2) stable;

  * Fix crash when serd_node_new_decimal is called with infinity or NaN
  * Fix crash when resolving against non-standard base URIs
  * Fix bug that caused "a" abbreviation in non-predicate position
  * Disable timestamps in HTML documentation for reproducible build
  * Fix clashing symbol "error" in amalgamation build
  * Update to waf 1.7.8 and autowaf r90 (install docs to versioned directory)

 -- David Robillard <d@drobilla.net>  Sat, 22 Dec 2012 21:32:15 -0500

serd (0.18.0) stable;

  * Support digits at start of local names as per new Turtle grammar
  * Add incremental read interface suitable for reading from infinite streams
  * Add -e option to serdi to use incremental reading
  * Add error callback to reader and writer for custom error reporting
  * Add -q option to serdi to suppress all non-data output, e.g. errors
  * Reset indent when finishing a write
  * Report write size correctly when invalid UTF-8 is encountered and a
    replacement character is written
  * Strip down API documentation to a single clean page
  * Fix various hyper-strict warnings
  * Do not require a C++ compiler to build
  * Add option to build utilities as static binaries
  * Upgrade to waf 1.7.2

 -- David Robillard <d@drobilla.net>  Thu, 23 Aug 2012 00:18:34 -0400

serd (0.14.0) stable;

  * Use path variables in pkgconfig files
  * Install man page to DATADIR (e.g. PREFIX/share/man, not PREFIX/man)

src/serdi.c
/*
  Copyright 2011-2017 David Robillard <http://drobilla.net>

  Permission to use, copy, modify, and/or distribute this software for any
  purpose with or without fee is hereby granted, provided that the above
  copyright notice and this permission notice appear in all copies.

  THIS SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*/

#include "serd_config.h"
#include "system.h"

#include "serd/serd.h"

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

#include <limits.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#define SERDI_ERROR(msg)       fprintf(stderr, "serdi: " msg);
#define SERDI_ERRORF(fmt, ...) fprintf(stderr, "serdi: " fmt, __VA_ARGS__);

static int
print_version(void)
{
	printf("serdi " SERD_VERSION " <http://drobilla.net/software/serd>\n");
	printf("Copyright 2011-2017 David Robillard <http://drobilla.net>.\n"
	       "License: <http://www.opensource.org/licenses/isc>\n"
	       "This is free software; you are free to change and redistribute it."
	       "\nThere is NO WARRANTY, to the extent permitted by law.\n");
	return 0;
}

static int
print_usage(const char* name, bool error)
{
	FILE* const os = error ? stderr : stdout;
	fprintf(os, "%s", error ? "\n" : "");
	fprintf(os, "Usage: %s [OPTION]... INPUT [BASE_URI]\n", name);
	fprintf(os, "Read and write RDF syntax.\n");
	fprintf(os, "Use - for INPUT to read from standard input.\n\n");
	fprintf(os, "  -a           Write ASCII output if possible.\n");
	fprintf(os, "  -b           Fast bulk output for large serialisations.\n");
	fprintf(os, "  -c PREFIX    Chop PREFIX from matching blank node IDs.\n");
	fprintf(os, "  -e           Eat input one character at a time.\n");
	fprintf(os, "  -f           Fast serialisation without inlining.\n");
	fprintf(os, "  -h           Display this help and exit.\n");
	fprintf(os, "  -i SYNTAX    Input syntax: turtle/ntriples/trig/nquads.\n");
	fprintf(os, "  -k BYTES     Parser stack size.\n");
	fprintf(os, "  -l           Lax (non-strict) parsing.\n");
	fprintf(os, "  -m           Build and serialise a model (no streaming).\n");
	fprintf(os, "  -o SYNTAX    Output syntax: turtle/ntriples/nquads.\n");
	fprintf(os, "  -p PREFIX    Add PREFIX to blank node IDs.\n");
	fprintf(os, "  -q           Suppress all output except data.\n");
	fprintf(os, "  -r ROOT_URI  Keep relative URIs within ROOT_URI.\n");
	fprintf(os, "  -s INPUT     Parse INPUT as string (terminates options).\n");
	fprintf(os, "  -t           Write terser output without newlines.\n");
	fprintf(os, "  -v           Display version information and exit.\n");
	return error ? 1 : 0;
}

static int
missing_arg(const char* name, char opt)
{
	SERDI_ERRORF("option requires an argument -- '%c'\n", opt);
	return print_usage(name, true);
}

static SerdStatus
quiet_error_func(void*               handle,
                 const char*         domain,
                 SerdLogLevel        level,
                 const SerdLogField* fields,
                 size_t              n_fields,
                 const char*         fmt,
                 va_list             args)
{
	(void)handle;
	(void)domain;
	(void)level;
	(void)fields;
	(void)n_fields;
	(void)fmt;
	(void)args;
	return SERD_SUCCESS;
}

int
main(int argc, char** argv)
{
	if (argc < 2) {
		return print_usage(argv[0], true);
	}

	SerdSyntax      input_syntax  = (SerdSyntax)0;
	SerdSyntax      output_syntax = (SerdSyntax)0;
	SerdWriterFlags writer_flags  = 0;
	bool            from_string   = false;
	bool            from_stdin    = false;
	bool            bulk_read     = true;
	bool            bulk_write    = false;
	bool            no_inline     = false;
	bool            lax           = false;
	bool            use_model     = false;
	bool            quiet         = false;
	size_t          stack_size    = 4194304;
	const char*     add_prefix    = NULL;
	const char*     chop_prefix   = NULL;
	const char*     root_uri      = NULL;
	int             a             = 1;
	for (; a < argc && argv[a][0] == '-'; ++a) {
		if (argv[a][1] == '\0') {
			from_stdin = true;
			break;
		} else if (argv[a][1] == 'a') {
			writer_flags |= SERD_WRITE_ASCII;
		} else if (argv[a][1] == 'b') {
			bulk_write = true;
		} else if (argv[a][1] == 'e') {
			bulk_read = false;
		} else if (argv[a][1] == 'f') {
			no_inline = true;
		} else if (argv[a][1] == 'h') {
			return print_usage(argv[0], false);
		} else if (argv[a][1] == 'l') {
			lax = true;
		} else if (argv[a][1] == 'm') {
			use_model = true;
		} else if (argv[a][1] == 'q') {
			quiet = true;
		} else if (argv[a][1] == 'v') {
			return print_version();
		} else if (argv[a][1] == 's') {
			from_string = true;
			++a;
			break;
		} else if (argv[a][1] == 't') {
			writer_flags |= SERD_WRITE_TERSE;
		} else if (argv[a][1] == 'i') {
			if (++a == argc) {
				return missing_arg(argv[0], 'i');
			} else if (!(input_syntax = serd_syntax_by_name(argv[a]))) {
				return print_usage(argv[0], true);
			}
		} else if (argv[a][1] == 'k') {
			if (++a == argc) {
				return missing_arg(argv[0], 'k');
			}
			char*      endptr = NULL;
			const long size   = strtol(argv[a], &endptr, 10);
			if (size <= 0 || size == LONG_MAX || *endptr != '\0') {
				SERDI_ERRORF("invalid stack size `%s'\n", argv[a]);
				return 1;
			}
			stack_size = (size_t)size;
		} else if (argv[a][1] == 'o') {
			if (++a == argc) {
				return missing_arg(argv[0], 'o');
			} else if (!(output_syntax = serd_syntax_by_name(argv[a]))) {
				return print_usage(argv[0], true);
			}
		} else if (argv[a][1] == 'p') {
			if (++a == argc) {
				return missing_arg(argv[0], 'p');
			}
			add_prefix = (const char*)argv[a];
		} else if (argv[a][1] == 'c') {
			if (++a == argc) {
				return missing_arg(argv[0], 'c');
			}
			chop_prefix = (const char*)argv[a];
		} else if (argv[a][1] == 'r') {
			if (++a == argc) {
				return missing_arg(argv[0], 'r');
			}
			root_uri = (const char*)argv[a];
		} else {
			SERDI_ERRORF("invalid option -- '%s'\n", argv[a] + 1);
			return print_usage(argv[0], true);
		}
	}

	if (a == argc) {
		SERDI_ERROR("missing input\n");
		return 1;
	}

#ifdef _WIN32
	_setmode(fileno(stdin), _O_BINARY);
	_setmode(fileno(stdout), _O_BINARY);
#endif

	const char* input = (const char*)argv[a++];

	if (!input_syntax && !(input_syntax = serd_guess_syntax(input))) {
		input_syntax = SERD_TRIG;
	}

	const bool input_has_graphs = serd_syntax_has_graphs(input_syntax);
	if (!output_syntax) {
		output_syntax = input_has_graphs ? SERD_NQUADS : SERD_NTRIPLES;
	}

	SerdNode* base = NULL;
	if (a < argc) {  // Base URI given on command line
		base = serd_new_uri((const char*)argv[a]);
	} else if (!from_string && !from_stdin) {  // Use input file URI
		base = serd_new_file_uri(input, NULL);
	}

	FILE*      out_fd = stdout;
	SerdWorld* world  = serd_world_new();
	SerdEnv*   env    = serd_env_new(base);

	const SerdSerialisationFlags serialisation_flags =
		no_inline ? SERD_NO_INLINE_OBJECTS : 0U;

	SerdByteSink* byte_sink = serd_byte_sink_new(
		(SerdWriteFunc)fwrite, out_fd, bulk_write ? 4096U : 1U);

	SerdWriter* writer = serd_writer_new(world,
	                                     output_syntax,
	                                     writer_flags,
	                                     env,
	                                     (SerdWriteFunc)serd_byte_sink_write,
	                                     byte_sink);

	SerdReader*     reader   = NULL;
	SerdModel*      model    = NULL;
	SerdInserter*   inserter = NULL;
	const SerdSink* sink     = NULL;
	if (use_model) {
		const SerdModelFlags flags =
		        SERD_INDEX_SPO | (input_has_graphs ? SERD_INDEX_GRAPHS : 0U) |
		        (no_inline ? 0U : SERD_INDEX_OPS);
		model    = serd_model_new(world, flags);
		inserter = serd_inserter_new(model, env, NULL);
		sink     = serd_inserter_get_sink(inserter);
	} else {
		sink = serd_writer_get_sink(writer);
	}

	reader = serd_reader_new(world, input_syntax, sink, stack_size);
	serd_reader_set_strict(reader, !lax);
	if (quiet) {
		serd_world_set_message_func(world, quiet_error_func, NULL);
	}

	SerdNode* root = serd_new_uri(root_uri);
	serd_writer_set_root_uri(writer, root);
	serd_writer_chop_blank_prefix(writer, chop_prefix);
	serd_reader_add_blank_prefix(reader, add_prefix);
	serd_node_free(root);

	SerdStatus status     = SERD_SUCCESS;
	SerdNode*  input_name = NULL;
	if (from_string) {
		input_name = serd_new_string("string");
		status = serd_reader_start_string(reader, input, input_name);
	} else if (from_stdin) {
		input_name = serd_new_string("stdin");
		status = serd_reader_start_stream(reader,
		                                  serd_file_read_byte,
		                                  (SerdStreamErrorFunc)ferror,
		                                  stdin,
		                                  input_name,
		                                  1);
	} else {
		status = serd_reader_start_file(reader, input, bulk_read);
	}

	if (!status) {
		status = serd_reader_read_document(reader);
	}

	serd_reader_finish(reader);

	if (status <= SERD_FAILURE && use_model) {
		const SerdSink* wsink = serd_writer_get_sink(writer);
		serd_env_write_prefixes(env, wsink);

		SerdRange* range = serd_model_all(model);
		status = serd_range_serialise(range, wsink, serialisation_flags);
		serd_range_free(range);
	}

	serd_node_free(input_name);
	serd_inserter_free(inserter);
	serd_model_free(model);
	serd_reader_free(reader);
	serd_writer_free(writer);
	serd_byte_sink_free(byte_sink);
	serd_env_free(env);
	serd_node_free(base);
	serd_world_free(world);

	if (fclose(stdout)) {
		perror("serdi: write error");
		status = SERD_ERR_UNKNOWN;
	}

	return (status > SERD_FAILURE) ? 1 : 0;
}
